app/controllers/step_results_controller.rb

```diff
@@ -1,53 +1,70 @@
-      render json: { step_result: { id: @step_result.id } }
+      @step_results.where.not(result: @results).destroy_all
+      @results.where.not(id: @step_results.select(:result_id)).each do |result|
+        StepResult.create!(step: @steps.first, result: result, created_by: current_user)
+      end
+      render json: { results: @steps.first.results.map { |r| { id: r.id, name: r.name } } }, status: :created
     rescue ActiveRecord::RecordInvalid => e
       Rails.logger.error e.message
-      render json: { errors: @step_result.errors.full_messages }, status: :unprocessable_entity
+      render json: { message: :error }, status: :unprocessable_entity
+      raise ActiveRecord::Rollback
     end
   end
 
-  def destroy
+  def link_steps
     ActiveRecord::Base.transaction do
-      if @step_result.destroy
-        render json: {}, status: :ok
-      else
-        render json: { errors: @step_result.errors.full_messages }, status: :unprocessable_entity
+      @step_results.where.not(step: @steps).destroy_all
+      @steps.where.not(id: @step_results.select(:step_id)).each do |step|
+        StepResult.create!(step: step, result: @results.first, created_by: current_user)
       end
+      render json: { steps: @results.first.steps.map { |s| { id: s.id, name: s.name } } }, status: :created
+    rescue ActiveRecord::RecordInvalid => e
+      Rails.logger.error e.message
+      render json: { message: :error }, status: :unprocessable_entity
+      raise ActiveRecord::Rollback
     end
-    @step_result.destroy!
   end
 
   private
 
-  def load_step
-    @step = Step.find_by(id: params[:step_id])
-    render_404 unless @step
-  end
+  def load_steps
+    @steps = Step.where(id: params[:step_ids])
+
+    render_404 and return if (action_name == 'link_results') && @steps.size != 1
 
-  def load_result
-    @result = Result.find_by(id: params[:result_id])
-    render_404 unless @result
+    render_403 and return if @steps.pluck(:protocol_id).uniq.size > 1
   end
 
-  def load_step_result
-    @step_result = StepResult.find_by(id: params[:id])
-    render_404 unless @step_result
+  def load_results
+    @results = Result.where(id: params[:result_ids])
 
-    @step = @step_result.step
-    @result = @step_result.result
+    render_404 and return if (action_name == 'link_steps') && @results.size != 1
+
+    render_403 and return if @results.pluck(:my_module_id).uniq.size > 1
+  end
+
+  def load_step_results
+    @step_results = StepResult.where(step: @steps) if action_name == 'link_results'
+    @step_results = StepResult.where(result: @results) if action_name == 'link_steps'
   end
 
   def check_manage_permissions
-    render_403 unless @step.my_module == @result.my_module && can_manage_my_module?(@step.my_module)
+    case action_name
+    when 'link_results'
+      render_403 and return unless can_manage_my_module?(@steps.first.my_module)
+    when 'link_steps'
+      render_403 and return unless can_manage_my_module?(@results.first.my_module)
```

Comment:
in these cases it should be in the form of
`@results.all? { |r| can_manage_my_module?(r.my_module) }`

Reply:
We checked previously that all results/steps belongs same my_module